### PR TITLE
Auto backend for pipeline and serve when backend is not set to pytorch explicitly

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support running new models supported by pytorch engine yet not by turbomind for pipeline and serve

## Modification

- update `lmdeploy.pipeline`, `lmdeploy.serve` to support auto backend
- update `api_server` and `gradio` cli to support auto backend

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/archs.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    # baichuan2-7b, baichuan-13b, baichuan2-13b
+    BaichuanForCausalLM=dict(pytorch=True, turbomind=True),
+    # chatglm2-6b, chatglm3-6b
+    ChatGLMModel=dict(pytorch=True, turbomind=False),
+    # deepseek-moe
+    DeepseekForCausalLM=dict(pytorch=True, turbomind=False),
+    # falcon-7b
+    FalconForCausalLM=dict(pytorch=True, turbomind=False),
+    # gemma-7b
+    GemmaForCausalLM=dict(pytorch=False, turbomind=False),
```
Comment (RunningLeon) [inline lmdeploy/archs.py:20]: need to update here after pr https://github.com/InternLM/lmdeploy/pull/1184

Comment (lvhan028) [reply lmdeploy/archs.py:20]: PR #1184 has been merged.

Comment (RunningLeon) [reply lmdeploy/archs.py:20]: ok

File: lmdeploy/model.py [hunk 2]
```diff
@@ -720,7 +720,7 @@ def get_prompt(self, prompt, sequence_start=True):
 @MODELS.register_module(name='chatglm2-6b')
 class ChatGLM2(BaseModel):
 
-    def __init__(self):
+    def __init__(self, **kwargs):
```
Comment (grimoire) [inline lmdeploy/model.py:723]: Do we need add kwargs in falcon?

File: lmdeploy/cli/cli.py [hunk 3]
```diff
@@ -107,7 +107,8 @@ def list(args):
         if engine == 'pytorch':
             model_names = [
                 'llama', 'llama2', 'internlm', 'internlm2', 'baichuan2',
-                'chatglm2', 'falcon', 'yi', 'mistral', 'qwen1.5', 'gemma'
+                'chatglm2', 'falcon', 'yi', 'mistral', 'mixtral', 'qwen1.5',
```
Comment (grimoire) [inline lmdeploy/cli/cli.py:110]: deepseek

File: lmdeploy/archs.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                - iii) The model_id of a model hosted inside a model repo
+                    on huggingface.co, such as "internlm/internlm-chat-7b",
+                    "Qwen/Qwen-7B-Chat ", "baichuan-inc/Baichuan2-7B-Chat"
+                    and so on.
+
+    Returns:
+        str: the backend type.
+    """
+    pytorch_has, turbomind_has = _is_support_by(model_path)
+    backend = 'turbomind' if turbomind_has else 'pytorch'
```
Comment (grimoire) [inline lmdeploy/archs.py:119]: Better warning if model has not been supported by both pytorch and turbomind.

File: lmdeploy/model.py (hunk 2 above)
Comment (RunningLeon) [reply lmdeploy/model.py:723]: it is fixed in https://github.com/InternLM/lmdeploy/pull/1168. maybe should remove it and wait for the pr

**Review by grimoire — APPROVED** [review]

File: lmdeploy/archs.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    # falcon-7b
+    FalconForCausalLM=dict(pytorch=True, turbomind=False),
+    # gemma-7b
+    GemmaForCausalLM=dict(pytorch=True, turbomind=False),
+    # internlm
+    InternLMForCausalLM=dict(pytorch=True, turbomind=True),
+    # internlm2
+    InternLM2ForCausalLM=dict(pytorch=True, turbomind=True),
+    # internlm-xcomposer
+    InternLM2XComposerForCausalLM=dict(pytorch=False, turbomind=True),
```
Comment (lvhan028) [inline lmdeploy/archs.py:29]: 这个两个引擎都不支持

File: lmdeploy/archs.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+from .messages import PytorchEngineConfig, TurbomindEngineConfig
+from .utils import get_logger
+
+logger = get_logger('lmdeploy')
+
+_SUPPORTED_ARCHS = dict(
+    # baichuan-7b
+    BaiChuanForCausalLM=dict(pytorch=False, turbomind=True),
+    # baichuan2-7b, baichuan-13b, baichuan2-13b
+    BaichuanForCausalLM=dict(pytorch=True, turbomind=True),
```
[the remark this answers is not in the document]
Comment (lvhan028) [reply lmdeploy/archs.py:15]: 哦，看到下面有特殊处理了

File: lmdeploy/archs.py (hunk 5 above)
Comment (RunningLeon) [reply lmdeploy/archs.py:29]: 好的

File: lmdeploy/model.py (hunk 2 above)
Comment (lvhan028) [reply lmdeploy/model.py:723]: #1168 has been merged

File: lmdeploy/archs.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Literal, Optional, Union
```
Comment (lvhan028) [inline lmdeploy/archs.py:2]: May add unit test cases

**Comment by lvhan028** [comment]
If user set turbomind config wtih `model_format==awq`, what will happen?

**Comment by RunningLeon** [comment]
> If user set turbomind config wtih `model_format==awq`, what will happen?

if the model is supported by turbomind, then using turbomind backend with awq. Else  use pytorch and `model_format` is ignored since pytorchengineconfig does not have `model_format`.

File: lmdeploy/turbomind/supported_models.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+from lmdeploy.utils import get_logger
+
+logger = get_logger('lmdeploy')
+
+_SUPPORTED_ARCHS = dict(
+    # baichuan-7b
+    BaiChuanForCausalLM=True,
+    # baichuan2-7b, baichuan-13b, baichuan2-13b
+    BaichuanForCausalLM=True,
```
Comment (AllentDan) [inline lmdeploy/turbomind/supported_models.py:12]: why two `BaiChuanForCausalLM`?

File: lmdeploy/archs.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+from .messages import PytorchEngineConfig, TurbomindEngineConfig
+from .utils import get_logger
+
+logger = get_logger('lmdeploy')
+
+
+def autoget_backend(model_path: str) -> Union[Literal['turbomind', 'pytorch']]:
```
Comment (AllentDan) [inline lmdeploy/archs.py:10]: model_path for this function. If it is a turbomind model, I mean it was converted. Then it should be turbomind.

File: lmdeploy/turbomind/supported_models.py (hunk 8 above)
Comment (RunningLeon) [reply lmdeploy/turbomind/supported_models.py:12]: they changed the name from `BaiChuanForCausalLM` to `BaichuanForCausalLM` after `baichuan-7b`.
https://huggingface.co/baichuan-inc/Baichuan-7B/blob/5d86e56a58fe4a5b3292cd9bb7468afef6f93eab/config.json#L3
https://huggingface.co/baichuan-inc/Baichuan-13B-Chat/blob/e3e1498bb6b7fb0bcb5a65679c09b862d7c29301/config.json#L5

**Review by irexyc — APPROVED** [review]

**Comment by lvhan028** [comment]
pytorch engine 支持 /workspace/models-140/baichuan/Baichuan-13B-Chat 吗？使用 pipeline接口测试：
```
AttributeError: 'BaichuanTokenizer' object has no attribute 'sp_model'
```

**Comment by lvhan028** [comment]
/workspace/models-140/baichuan2/Baichuan2-13B-Chat，
```
AssertionError('Loop-carried variable acc has initial type <[32], bf16> but is re-assigned to <[32], fp32> in loop! Please make sure that the type stays consistent.')
2024-03-01 12:04:00,632 - lmdeploy - ERROR - Engine main loop stopped.
```

**Comment by lvhan028** [comment]
/workspace/models-142/models/falcon-7b-instruct/

```
AssertionError('Loop-carried variable acc has initial type <[32], bf16> but is re-assigned to <[32], fp32> in loop! Please make sure that the type stays consistent.')
```

**Comment by RunningLeon** [comment]
> workspace/model
@lvhan028 this fails on main branch too. Maybe not related with changes in this PR.

**Comment by RunningLeon** [comment]
> pytorch engine 支持 /workspace/models-140/baichuan/Baichuan-13B-Chat 吗？使用 pipeline接口测试：
> 
> ```
> AttributeError: 'BaichuanTokenizer' object has no attribute 'sp_model'
> ```

Baichuan-13B-Chat is not explicitly supported by turbomind and pytorch. So try to run with pytorch as the warning says:
```
2024-03-01 20:13:53,684 - lmdeploy - WARNING - /share/140/baichuan/Baichuan-13B-Chat is not explicitly supported by lmdeploy. Try to run with lmdeploy pytorch engine.
```

**Comment by AllentDan** [comment]
> pytorch engine 支持 /workspace/models-140/baichuan/Baichuan-13B-Chat 吗？使用 pipeline接口测试：
> 
> ```
> AttributeError: 'BaichuanTokenizer' object has no attribute 'sp_model'
> ```

Transformers 版本升级后，baichuan模型没有跟进。感觉可以不用管

**Comment by grimoire** [comment]
> > workspace/model
> 
> @lvhan028 this fails on main branch too. Maybe not related with changes in this PR. @grimoire any idea on this issue?

No idea.

**Comment by grimoire** [comment]
> pytorch engine 支持 /workspace/models-140/baichuan/Baichuan-13B-Chat 吗？使用 pipeline接口测试：
> 
> ```
> AttributeError: 'BaichuanTokenizer' object has no attribute 'sp_model'
> ```

没支持 baichuan1

**Comment by grimoire** [comment]
> > workspace/model
> 
> @lvhan028 this fails on main branch too. Maybe not related with changes in this PR. @grimoire any idea on this issue?

https://github.com/InternLM/lmdeploy/pull/1228

**Review by AllentDan — APPROVED** [review]
api_server tested OK with model `internlm-chat-20b` and `llama2-chat-7b` for both the backends.

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**